// ardeidae/lib/ardeidaeMsgControl.js

/**
 * Message object
 */
function Message(userId, message) {
  var aaa = UsrControl.findNameByIndex(userId),
        bbb = UsrControl.findOriginByIndex(userId);
  this.time = Date.now();
  this.userId = userId;
  this.userName = aaa !== undefined ? aaa : 'server';
  // this.userName = UserList[userId].name || 'server';
  this.userName = bbb !== undefined ? bbb : 'server';
  // this.origin = UserList[userId].origin;
  this.message = message;
}


/**
 * Message Controller object
 */
function MessageController () {
  this.MessageLog = [];            // Log for users messages
  this.ServerMessageLog = [];
}
MessageController.prototype = {
  prepareEcho: function (senderName, msg) {
    var obj = {name: senderName,
                    message: htmlEntities( msg ),
                  };
    var portable = JSON.stringify(obj);
    return portable;
  },

  preparePrivateEcho: function (senderName, msg) {
    var obj = {name: senderName,
                    message: htmlEntities( msg ),
                    attributes: "private" };
    var portable = JSON.stringify(obj);
    return portable;
  },

  usersToString: function () {
      var i, string = '';
      for( i = 0; i < this.UserList.length; i++ ) {
        if (this.UserList[i]) {
          string += ',' + this.UserList[i].name;
        }
      }
      return string;
  },


};




module.exports = MessageController;
