// ardeidae/lib/ardeidaeUsrControl.js
/**
 * User object
 */
function User(id, origin) {
  this.time = Date.now();
  this.id = id;
  this.origin = origin || 'remote user';
  this.name = 'no-name';
}


/**
 * User Controller object
 */
function UserController() {
  this.UserList = [];
  this.ActiveUsers = null;
}
UserController.prototype = {
  usersToString: function () {
      var i, string = '';
      for( i = 0; i < this.UserList.length; i++ ) {
        if (this.UserList[i]) {
          string += ',' + this.UserList[i].name;
        }
      }
      return string;
  },

  printUserListArray: function () {
    console.log('No. of Online users: ' + this.ActiveUsers + ', Length of UserList array: ' + this.UserList.length);
    var i;
    for ( i = 0; i < this.UserList.length; i++ ) {
      if ( this.UserList[i] ) {
        console.log('ID: ' + this.UserList[i].id + ', at index: ' + i  + ', from: ' + this.UserList[i].origin + ', created: ' + this.UserList[i].time + ', name: ' + this.UserList[i].name);
      }
    }
  },

  getStats: function () {
    var feedback = 'stat,' + this.ActiveUsers + this.usersToString();
    return feedback;
  },

  getHistory: function () {
    return 'No data available';
  },

  getArrayLength: function () {
    return this.UserList.length;
  },

  userCountUp: function () {
    this.ActiveUsers++;
  },

  userCountDown: function () {
    this.ActiveUsers--;
  },

  addNewUser: function (ID, remoteAddress) {
    this.userCountUp();
    var newUserId = ID;
    console.log(Date.now() + ': Creating new user with id: ' + newUserId + ' From origin: ' + remoteAddress);
    this.UserList.push( new User(newUserId, remoteAddress) );
  },

  removeByIndex: function (index) {
      this.UserList[index] = null;
      this.userCountDown();
  },

  findIndexByName: function (name) {
    var i;
    for( i = 0; i < this.UserList.length; i++ ) {
        if ( this.UserList[i].name === name ) {
          return i;
        }
    }
  },

  findNameByIndex: function (index) {
    if (this.UserList[index]) {
        return this.UserList[index].name;
    }
    return null;
  },

  findOriginByIndex: function (index) {
        if (this.UserList[index]) {
        return this.UserList[index].origin;
    }
    return null;
  },

  findIndexByRemoteAddress: function (address) {
    var i;
    for( i = 0; i < this.UserList.length; i++ ) {
      console.log('Searching db for remote addresses, found: ' + this.UserList[i].origin + ' at index: ' + i);
        if ( this.UserList[i].origin === address ) {
          return i;
        }
    }
  },

  setNameAtIndex: function (name, index) {
    if (this.UserList[index]) {
      this.UserList[index].name = name;
    }
  }
};

module.exports = UserController;
